src/QsCompiler/SyntaxProcessor/ExpressionVerification.fs

```diff
@@ -3,801 +3,300 @@
+/// </summary>
+/// <returns>The intersection type and the diagnostics.</returns>
+let private VerifyConcatenation (inference: InferenceContext) range lhs rhs =
+    let exType, intersectDiagnostics = inference.Intersect(lhs.ResolvedType, rhs.ResolvedType)
+    let exType = exType |> ResolvedType.withAllRanges (TypeRange.inferred range)
+    let constrainDiagnostics = inference.Constrain(exType, Semigroup)
+    exType, intersectDiagnostics @ constrainDiagnostics
+
+/// <summary>
+/// Verifies that <paramref name="lhs"/> and <paramref name="rhs"/> have an intersecting equatable type.
+/// </summary>
+/// <returns>The intersection type and the diagnostics.</returns>
+let private VerifyEqualityComparison (inference: InferenceContext) range lhs rhs =
+    let exType, intersectDiagnostics = inference.Intersect(lhs.ResolvedType, rhs.ResolvedType)
+    let exType = exType |> ResolvedType.withAllRanges (TypeRange.inferred range)
+    let constrainDiagnostics = inference.Constrain(exType, Equatable)
+    intersectDiagnostics @ constrainDiagnostics
+
+/// <summary>
+/// Verifies that <paramref name="exprs"/> can form an array.
+/// </summary>
+/// <returns>The type of the array and the diagnostics.</returns>
+let private VerifyValueArray (inference: InferenceContext) range exprs =
+    let diagnostics = ResizeArray()
+
+    for expr in exprs do
+        if expr.ResolvedType.isMissing then
+            QsCompilerDiagnostic.Error (ErrorCode.MissingExprInArray, []) (rangeOrDefault expr)
+            |> diagnostics.Add
+
+    let types =
+        exprs
+        |> Seq.map (fun e -> e.ResolvedType)
+        |> Seq.filter (fun t -> not t.isInvalid && not t.isMissing)
+
+    if Seq.isEmpty types && Seq.isEmpty exprs then
+        inference.Fresh range |> ArrayType |> ResolvedType.create (Inferred range), Seq.toList diagnostics
+    elif Seq.isEmpty types then
+        ResolvedType.New InvalidType |> ArrayType |> ResolvedType.create (Inferred range), Seq.toList diagnostics
+    else
+        types
+        |> Seq.reduce (fun left right ->
+            let intersectionType, intersectionDiagnostics = inference.Intersect(left, right)
+            List.iter diagnostics.Add intersectionDiagnostics
+            intersectionType |> ResolvedType.withAllRanges right.Range)
+        |> ResolvedType.withAllRanges (Inferred range)
+        |> ArrayType
+        |> ResolvedType.create (Inferred range),
+        Seq.toList diagnostics
+
+/// <summary>
+/// Verifies that <paramref name="expr"/> has a type that can be indexed by an expression of type Range.
+/// </summary>
+/// <returns>The item type and the diagnostics.</returns>
+let internal VerifyNumberedItemAccess (inference: InferenceContext) expr =
+    let range = rangeOrDefault expr
+    let itemType = inference.Fresh range
+    itemType, inference.Constrain(expr.ResolvedType, Indexed(ResolvedType.New Range, itemType))
```

Comment:
I changed this function so that it is hopefully less confusing. See this comment: https://github.com/microsoft/qsharp-compiler/pull/940#discussion_r608232778